libraries/USBSerial/src/usbd_cdc_if.c

```diff
@@ -0,0 +1,448 @@
+  if ((UserRxBufPtrOut + APP_RX_DATA_SIZE - UserRxBufPtrIn - 1) %
+              APP_RX_DATA_SIZE +
+          1 >=
+      USB_OTG_FS_MAX_PACKET_SIZE)
+    USBD_CDC_ReceivePacket(
+        &hUSBD_Device_CDC); // Initiate next USB packet transfer once a packet
+                            // is received and there is enouch space in the
+                            // buffer
+  else
+    receiveSuspended = true;
+
+  return (USBD_OK);
+}
+
+/**
+ * @brief  CDC_Transmit_FS
+ *         Data send over USB IN endpoint are sent over CDC interface
+ *         through this function.
+ *         @note
+ *
+ *
+ * @param  Buf: Buffer of data to be send
+ * @param  Len: Number of data to be send (in bytes)
+ * @retval Result of the operation: USBD_OK if all operations are OK else
+ * USBD_FAIL or USBD_BUSY
+ */
+uint8_t CDC_Transmit_FS(uint8_t *Buf, uint16_t Len) {
+  UNUSED(Buf);
+  uint8_t result = USBD_OK;
+  UNUSED(Len);
+  USBD_CDC_SetTxBuffer(&hUSBD_Device_CDC, (uint8_t *)UserTxBufferFS, 1);
+  result = USBD_CDC_TransmitPacket(&hUSBD_Device_CDC);
+  return result;
+}
+
+void CDC_flush(void) {
+  uint8_t status;
+
+  if (UserTxBufPtrOut != UserTxBufPtrIn) {
+    if (UserTxBufPtrOut > UserTxBufPtrIn) /* Roll-back */
+    {
+      memcpy((uint8_t *)&StackTxBufferFS[0],
+             (uint8_t *)&UserTxBufferFS[UserTxBufPtrOut],
+             (APP_TX_DATA_SIZE - UserTxBufPtrOut));
+      memcpy((uint8_t *)&StackTxBufferFS[APP_TX_DATA_SIZE - UserTxBufPtrOut],
+             (uint8_t *)&UserTxBufferFS[0], UserTxBufPtrIn);
+
+      USBD_CDC_SetTxBuffer(
+          &hUSBD_Device_CDC, (uint8_t *)&StackTxBufferFS[0],
+          (APP_TX_DATA_SIZE - UserTxBufPtrOut + UserTxBufPtrIn));
+    } else {
+      USBD_CDC_SetTxBuffer(&hUSBD_Device_CDC,
+                           (uint8_t *)&UserTxBufferFS[UserTxBufPtrOut],
+                           (UserTxBufPtrIn - UserTxBufPtrOut));
+    }
+
+    do {
+      if (lineState == 0) // Device disconnected
+        status = USBD_OK;
+      else
+        status = USBD_CDC_TransmitPacket(&hUSBD_Device_CDC);
+    } while (status == USBD_BUSY);
+
+    if (status == USBD_OK) {
+      UserTxBufPtrOut = UserTxBufPtrIn;
+    }
+  }
+}
+
+void CDC_disable_TIM_Interrupt(void) { HAL_NVIC_DisableIRQ(TIM6_DAC_IRQn); }
+
+void CDC_enable_TIM_Interrupt(void) { HAL_NVIC_EnableIRQ(TIM6_DAC_IRQn); }
+
+void CDC_resume_receive(void) {
+  if (receiveSuspended) {
+    if ((UserRxBufPtrOut + APP_RX_DATA_SIZE - UserRxBufPtrIn - 1) %
+                APP_RX_DATA_SIZE +
+            1 >=
+        USB_OTG_FS_MAX_PACKET_SIZE) {
+      USBD_CDC_ReceivePacket(
+          &hUSBD_Device_CDC); // Initiate next USB packet transfer once a packet
+                              // is received and there is enouch space in the
+                              // buffer
+      receiveSuspended = false;
+      HAL_GPIO_WritePin(GPIOD, GPIO_PIN_10, GPIO_PIN_RESET);
+    }
+  }
+}
+
+static void TIM_Config(void) {
+  /* Set TIMx instance */
+  TimHandle.timer = TIM6;
```

Comment:
I'm not certain this works across hardware. On 103CB I'm getting TIM6 doesn't exist errors, probably because it doesn't. Should this be generalized (or better yet, configured with a sane default for generic hardware?

Reply:
All Cortex-M cores have SysTick, so maybe using that would be the most platform independent way?

Reply:
This is now configurable. See #388